sui_programmability/examples/games_with_chance/sources/satoshi_flip.move

```diff
@@ -0,0 +1,304 @@
+    }
+
+    /// On an ended game (that has a non null outcome value) get what did the player guess
+    public fun guess(game: &mut Game): u8 {
+        assert!(option::is_some<Outcome>(&game.outcome), EGameNotEnded);
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.guess
+    }
+
+    /// Called by the player for a game with null BetData
+    public entry fun bet(game: &mut Game, guess: u8, stake_coin: Coin<SUI>, stake_amount: u64, ctx: &mut TxContext) {
+        assert!(option::is_none(&game.bet_data), EAlreadyAcceptedBet);
+        let house_data = option::borrow(&game.house_data);
+        assert!(stake_amount <= house_data.max_bet, EStakeTooHigh);
+        assert!(stake_amount > house_data.min_bet, EStakeTooLow);
+        assert!(guess == 0 || guess == 1, EGuessNot1Or0);
+        assert!(coin::value(&stake_coin) >= stake_amount, EPlayerCoinNotEnoughBalance);
+        game.epoch = tx_context::epoch(ctx);
+        // Get a balance with the stake_amount value
+        let stake = coin::into_balance(stake_coin);
+        if (balance::value(&stake) > stake_amount) {
+            let total_balance = balance::value(&stake);
+            let to_return = balance::split(&mut stake, total_balance - stake_amount);
+            // return the rest
+            transfer::transfer(coin::from_balance(to_return, ctx), tx_context::sender(ctx));
+        };
+        let bet_data = BetData {
+            stake: stake,
+            guess
+        };
+        option::fill(&mut game.bet_data, bet_data);
+        option::fill(&mut game.player, tx_context::sender(ctx));
+    }
+
+    /// Called by the house either after `fun bet` has been called by some player to end the game,
+    /// either before bet was called to "cancel" the game.
+    ///
+    /// If the house has "forgotten" the secret, then it can use a random secret and it will
+    /// cancel the game, as long as `fun bet` has not been called.
+    public entry fun end_game(game: &mut Game, secret: vector<u8>, ctx: &mut TxContext) {
+        // only house should be able to end the game
+        assert!(game.house == tx_context::sender(ctx), ENotAllowedToEndGame);
+
+        // house wants to cancel the current game (maybe forgot the secret)
+        if (option::is_none(& game.bet_data)) {
+            let msg = b"Game Canceled by House";
+            let outcome = Outcome {
+                secret: msg,
+                guess: 2,
+                player_won: false
+            };
+            option::fill(&mut game.outcome, outcome);
+
+            // no bet placed, return the balance to house
+            let HouseData {house_balance, min_bet: _, max_bet: _} = option::extract(&mut game.house_data);
+            let house_coins = coin::from_balance(house_balance, ctx);
+            transfer::transfer(house_coins, game.house);
+        } else {
+            let hash = sha3_256(secret);
+            if (hash != digest::sha3_256_digest_to_bytes(&game.hashed_secret)) {
```

Comment:
the house has no incentive to send an invalid hash even if we don't punish anyone, no?
if so, we can just use assert here